src/logger/mongoose_json_formatter.erl

```diff
@@ -11,28 +11,62 @@
 -export([format/2]).
 
 -spec format(logger:log_event(), logger:formatter_config()) -> unicode:chardata().
-format(E = #{msg := {report, #{label := {error_logger, _}, format := Format, args := Terms}}}, FConfig) ->
-    format(E#{msg := {report,
+format(Event, FConfig) ->
+    try do_format(Event, FConfig)
+    catch
+        %% Errors during log formatting can lead to a death spiral of recursive error logging, so
+        %% format the formatter error in a safe way and don't allow the exception to propagate.
+        error:Reason:Stacktrace -> format_log_formatter_error(error, Reason, Stacktrace, Event, FConfig)
+    end.
+
+
+format_log_formatter_error(Class, Reason, Stacktrace, #{meta := #{time := Time}} = Event, FConfig) ->
+    IoData = jiffy:encode(
+        #{
+            'when' => format_time(Time),
+            level => error,
+            what => log_format_failed,
+            class => Class,
+            reason => Reason,
+            stacktrace => format_item(Stacktrace, FConfig),
+            formatter_module => ?MODULE,
+            original_event => unicode:characters_to_binary(io_lib:format("~0p", [Event]))
+        }
+    ),
+    append_lf(IoData).
+
+
+-spec do_format(logger:log_event(), logger:formatter_config()) -> unicode:chardata().
+do_format(E = #{msg := {report, #{label := {error_logger, _}, format := Format, args := Terms}}}, FConfig) ->
+    do_format(E#{msg := {report,
                       #{unstructured_log =>
                         unicode:characters_to_binary(io_lib:format(Format, Terms))}}},
            FConfig);
-format(E = #{msg := {string, String}}, FConfig) ->
-    format(E#{msg := {report,
+do_format(E = #{msg := {string, String}}, FConfig) ->
+    do_format(E#{msg := {report,
                       #{unstructured_log =>
                         unicode:characters_to_binary(io_lib:format(String, []))}}}, FConfig);
-format(E = #{msg := {report, Report}}, FConfig) when is_map(Report) ->
+do_format(E = #{msg := {report, Report}}, FConfig) when is_map(Report) ->
     NewMap = process_metadata(E),
     NewReport = maps:merge(Report, NewMap),
     NewConfig = maps:merge(default_config(), config_correct_depth(FConfig)),
     Formatted = format_item(NewReport, NewConfig),
-    B = jiffy:encode(Formatted),
-    <<B/binary, "\n">>;
-format(Map = #{msg := {Format, Terms}}, FConfig) ->
-    format(Map#{msg := {report,
+    IoData = jiffy:encode(Formatted),
+    append_lf(IoData);
+do_format(Map = #{msg := {Format, Terms}}, FConfig) ->
+    do_format(Map#{msg := {report,
                         #{unstructured_log =>
                           unicode:characters_to_binary(io_lib:format(Format, Terms))}}},
            FConfig).
 
+
+-spec append_lf(iodata()) -> iodata().
+append_lf(IoData) when is_list(IoData) ->
```

Comment:
just this is enough:

```
append_lf(IoData) ->
    [IoData, <<"\n">>].
```

I am dialyzer:

```
-spec format(logger:log_event(), logger:formatter_config()) -> unicode:chardata().
-spec append_lf(iodata()) -> iodata().
```

These two return different types.

Log formatters return `unicode:chardata()`.

Should be:

```
-spec append_lf(unicode:chardata()) -> unicode:chardata().
```

(tldr: iodata - integers are bytes, unicode - untegers are code points).

There is no problem with the code, there is a problem with a spec :)

Reply:
I've made the change - got rid of the function because it didn't seem worth having it.  That means the spec has gone as well.

The old code always returned the `iodata()` from `jiffy:encode` as the `unicode:chardata()` return value of `format`.   If I understand the typing correctly, `iodata()` is the same as `unicode:chardata()` except when it's an (improper) list containing integers >=128, however there won't be any of those because the list returned from `jiffy:encode` consists of UTF-8 binaries.  So the semantic difference between `iodata()` and `unicode:chardata()` isn't a problem in this particular case (I've resisted the urge to write a code comment to this effect).

Reply:
Right, the only difference between unicode and iodata is when lists of integers are used.
Still, it's a good idea to avoid `iodata()` type in the unicode code, because currently it's fine, but after 3-4 generations of code modifications, there could be a point, when someone would add `iolist_to_binary(X)` and another person would add `io_lib:format("~ts", [Info]) (which returns code points higher 255)` in two parts of the code :D

Anyway, awesome contribution :)